runtime/lua/lsp/autocmds.lua

```diff
@@ -0,0 +1,225 @@
+  end
+
+  doautocmd(string.format('User %s/%s/%s', AutocmdEnum.AuPrefix, method_name, stage_name))
+end
+
+--- Get the event string
+-- @param autocmd_item (string|table)   - The autocmd event that will trigger this autocmd
+-- @param autocmd_pattern (string)      - (Optional) See |autocmd-patterns|. If not specified, '<buffer>'
+local get_autocmd_event_name = function(autocmd_item, autocmd_pattern)
+  if autocmd_pattern == nil or autocmd_pattern == '' then
+    autocmd_pattern = '<buffer>'
+  end
+
+  local autocmd_string
+  if type(autocmd_item) == 'string' then
+    autocmd_string = string.format('%s %s', autocmd_item, autocmd_pattern)
+  elseif type(autocmd_item == 'table') then
+    -- This is typically {'User', 'LSP/...'} -> 'User LSP/*'
+
+    -- Since User autocmds these are not registered exclusively to the buffer, we should only register them once
+    if autocmd_item[1] == 'User' then
+      if initialized_autocmds[autocmd_item] then
+        return ''
+      end
+
+      initialized_autocmds[autocmd_item] = true
+    end
+
+    autocmd_string = table.concat(autocmd_item, ' ')
+  else
+    log.debug('Unknown autocmd_item type: ', autocmd_item)
+    return ''
+  end
+
+  return autocmd_string
+end
+
+--- Register an autocmd with neovim
+-- @param request_name (string)         - Name of the request to register
+-- @param autocmd_item (string|table)   - The autocmd event that will trigger this autocmd
+-- @param autocmd_pattern (string)      - (Optional) See |autocmd-patterns|. If not specified, '*'
+local nvim_enable_autocmd = function(request_name, autocmd_item, autocmd_pattern)
+  local autocmd_event = get_autocmd_event_name(autocmd_item, autocmd_pattern)
+
+  if #autocmd_event == 0 then
+    return
+  end
+
+  local silent_level = 'silent!'
+  local command = string.format(
+    [[%s autocmd %s lua require('lsp.plugin').client.request_async('%s')]],
+    silent_level,
+    autocmd_event,
+    request_name
+  )
+
+  vim.api.nvim_command(command)
+
+  -- TODO: Determine if we need to be adding these to the "default autocmds"
+  -- if util.table.is_empty(default_autocmds[request_name]) then
+  --   return
+  -- end
+  -- if not util.table.is_empty(default_autocmds[request_name]) then
+  --   table.insert(default_autocmds[request_name], autocmd_item)
+  -- end
+end
+
+--- Export the autocmds from the table
+-- @param autocmd_table (table)     - (Optional) Table to give the list of autocmds to generate.
+--                                      If not passed in, then we will use the default tables.
+-- @param autocmd_pattern (string)  - (Optional) See |autocmd-patterns|. If not specified, '<buffer>'
+local export_autocmds = function(autocmd_table, autocmd_pattern)
+  if util.table.is_empty(autocmd_table) then
+    autocmd_table = default_autocmds
+  end
+
+  for request_name, autocmd_list in pairs(autocmd_table) do
+    for _, autocmd_item in ipairs(autocmd_list) do
+      print(request_name, autocmd_item, autocmd_pattern)
```

Comment:
I guess this print statement should be deleted, I get messages about autocmds when I switch from one .c buffer to another.